### PR TITLE
tracing: add support for the DataDog open-tracer

### PR DESCRIPTION
**Pull request by andreimatei**
Release note: Added support for the DataDog tracer. Set the
"trace.datadog.agent" cluster setting to enable it if you've got the
DataDog collector service running.

### CODE REVIEW

File: pkg/util/tracing/tracer.go [hunk 1]
```diff
@@ -90,6 +90,18 @@ var ZipkinCollector = settings.RegisterStringSetting(
 	envutil.EnvOrDefaultString("COCKROACH_TEST_ZIPKIN_COLLECTOR", ""),
 ).WithPublic()
 
+var dataDogAgentAddr = settings.RegisterStringSetting(
+	"trace.datadog.agent",
+	"if set, traces will be sent to this DataDog agent; use <host>:<port> or \"default\" for localhost:8126",
```
Comment (tbg) [inline pkg/util/tracing/tracer.go:95]: We only support one shadow manager, so every cluster setting will have to contain a caveat that it mutually conflicts with any of the other ones. Any appetite for fixing that limitation and unifying net/trace, lightstep, zipkin, dd, etc?

Also, can this be tested? See https://github.com/cockroachdb/cockroach/blob/2b71afb64fc7f8436085f1e6bf3b2b9bdc4f5364/pkg/util/tracing/tracer_test.go#L330

**Review by tbg — COMMENTED** [review]
I'm curious what prompted the introduction of this, and whether you could add that to the commit msg.

**Review by andreimatei — COMMENTED** [review]
I don't really have anything to add; it's at face value. And besides the face value, I went through some dependency hell for a bit trying to wrestle with modules and stuff. So we should save the next fool the trouble.
It was prompted by me wanting a refresher into what the state of the art in trace collectors is out there - hence the lightstep update and this. And DataDog I think is the dominant guy these days; we definitely have users using it. And CC uses it too. This integration works, except that I haven't been able to see log messages in the collected spans. Haven't dug yet.

FWIW, with both Lightstep and DD, it seems to me that the integration is not useful enough for the average user to recommend it very broadly (not to mention perf impact).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61602#-:-MVSRLvB-EAlGSOnbxJi:bim05yf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @tbg)

___
*[pkg/util/tracing/tracer.go, line 95 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61602#-MVEs3NeANrIKc7xVycT-r1-95:-MVSRB2I8kcO75dijuk1:b9jsckz) ([raw file](https://github.com/cockroachdb/cockroach/blob/421cfa375e3535da47efc89aa41041bba4d4f526/pkg/util/tracing/tracer.go#L95)):*
> We only support one shadow manager, so every cluster setting will have to contain a caveat that it mutually conflicts with any of the other ones. Any appetite for fixing that limitation and unifying net/trace, lightstep, zipkin, dd, etc?

I was slightly dissappointed by that myself. But not much appetite at the moment tbh, no.

> Also, can this be tested? See

Done

**Review by tbg — APPROVED** [review]
> Reviewed 7 of 7 files at r1, 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61602)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @irfansharif, and @tbg)

___
*[pkg/util/tracing/tracer.go, line 101 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61602#-MVVU7nbDpSa7VLOD7SK:-MVVU7nbDpSa7VLOD7SL:b-x2nzem) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae81b310c581baa52aaa42e147fa3882dd15e060/pkg/util/tracing/tracer.go#L101)):*
> ```Go
> var dataDogProjectName = settings.RegisterStringSetting(
> 	"trace.datadog.project",
> 	"the project under which traces will be reported to the DataDog agent if trace.datadog.agent is set",
> ```

Go through all of the shadow tracer settings and add a sentence "Only one of trace.datadog.project, ..., ... will take effect."

**Review by andreimatei — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61602)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @tbg)

___
*[pkg/util/tracing/tracer.go, line 101 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/61602#-MVVU7nbDpSa7VLOD7SK:-MVbAEbd9T_DFAJW1QS5:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae81b310c581baa52aaa42e147fa3882dd15e060/pkg/util/tracing/tracer.go#L101)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Go through all of the shadow tracer settings and add a sentence "Only one of trace.datadog.project, ..., ... will take effect."
</blockquote></details>

done

**Review by andreimatei — COMMENTED** [review]
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61602#-:-MVbJjAb4b561Pf5h3Y2:b-ydi9ay)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif and @tbg)

**Pull request merged**